DROOLS-3143 [DMN Designer] DMN unmarshal from 1.1, 1.2; marshal as 1.2

tarilabs opened the pull request:
I have not yet renamed package which hold the old `v1_1` reference in the hope to simplify the review, we can either leave it unchanged or refactor the package rename later once we agree on the actual code ;)

/cc @jomarko @manstis @etirelli 

please notice this requires https://github.com/kiegroup/drools/pull/2116

tarilabs commented:
Jenkins test again

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -173,98 +178,98 @@ public Graph unmarshall(final Metadata metadata,
                         String reqInputID = getId(ir.getRequiredInput());
                         Node requiredNode = elems.get(reqInputID).getValue();
-                        Edge myEdge = factoryManager.newElement(UUID.uuid(),
+                        Edge myEdge = factoryManager.newElement(ir.getId() != null ? ir.getId() : UUID.uuid(),
```
I've noticed that this condition happens many times in this class, maybe we could extract this to something like:
```java
private String getNonNullUUID(final String id) {
    return Optional.ofNullable(id).orElse(UUID.uuid());
}
```
..and re-use this in the lines 181, 191, 202, 212, 224, 234, 247, 257, 267 and 293.

tarilabs replied:
OK.
Mentioning for repeated should be just fine ;)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -366,17 +371,41 @@ private void connectEdge(final Edge edge,
+            Point target = e.getWaypoint().get(e.getWaypoint().size() - 1);
+            connectionContent.setTargetConnection(MagnetConnection.Builder.at(target.getX(), target.getY()));
+            if (e.getWaypoint().size() > 2) {
+                List<Point> sublist = e.getWaypoint().subList(1, e.getWaypoint().size() - 2);
```
I think we could cover this branch with tests.

tarilabs replied:
ok

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -410,31 +444,57 @@ public String marshall(final Diagram<Graph, Metadata> diagram) throws IOExceptio
+                            dmnEdge.setDmnElementRef(new QName(e.getUUID()));
+                            dmnEdge.getWaypoint().add(PointUtils.point2dToDMNDIPoint(sourceConn.getLocation()));
+                            for (ControlPoint cp : connectionContent.getControlPoints()) {
+                                dmnEdge.getWaypoint().add(PointUtils.point2dToDMNDIPoint(cp.getLocation()));
```
I think we could cover this line by tests too.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -537,14 +633,15 @@ private static void applyBounds(final RectangleDimensionsSet dimensionsSet,
     private static void applyBackgroundStyles(final BackgroundSet bgset,
                                               final DMNShape result) {
+        if (!(result.getStyle() instanceof DMNStyle)) {
+            return;
```
This logic is not covered by tests.

tarilabs replied:
It is not indeed, because by DMN XSD design this will not happen just by making use of the standard without extensions.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -473,59 +533,95 @@ private void internalAugment(Stream<DMNShape> drgShapeStream, Id id, Bound ul, R
 
+    private static void applyFontStyle(FontSet fontSet, DMNShape result) {
+        if (!(result.getStyle() instanceof DMNStyle)) {
+            return;
```
This logic is not covered by tests.

tarilabs replied:
ref reply

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java:
```diff
@@ -304,23 +302,28 @@ public void testLoan() throws IOException {
     }
 
+    @Test
+    public void temp() {
```
I think we can remove this method, right? 🙃

tarilabs replied:
Oops thanks for spotting! :)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java:
```diff
@@ -342,49 +345,45 @@ public void test_diamond() throws IOException {
-        assertDMNStyle("Open Sans", 24, 1, 255, 0, 0, myname.getFontStyle());
+        assertColor(255, 255, 255, ((DMNStyle) myname.getStyle()).getFillColor());
+        assertColor(0, 0, 0,       ((DMNStyle) myname.getStyle()).getStrokeColor());
+        //        assertEquals(0.5, myname.getBorderSize().getValue(), 0);
```
I think we could remove this line..

tarilabs replied:
Oops yes, as mentioned DMNDI does not support for border so I am not persisting this data.

tarilabs replied:
(mentioning repeated should fine ;) )

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java:
```diff
@@ -342,49 +345,45 @@ public void test_diamond() throws IOException {
-        assertDMNStyle("Times New Roman", 8, 2.5, 70, 60, 50, prefix.getFontStyle());
+        assertColor(0, 253, 25, ((DMNStyle) prefix.getStyle()).getFillColor());
+        assertColor(253, 0, 0, ((DMNStyle) prefix.getStyle()).getStrokeColor());
+        //        assertEquals(1, prefix.getBorderSize().getValue(), 0);
```
..and this one

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java:
```diff
@@ -342,49 +345,45 @@ public void test_diamond() throws IOException {
-        assertDMNStyle("Arial", 10, 1.5, 50, 60, 70, postfix.getFontStyle());
+        assertColor(247, 255, 0, ((DMNStyle) postfix.getStyle()).getFillColor());
+        assertColor(0, 51, 255, ((DMNStyle) postfix.getStyle()).getStrokeColor());
+        //        assertEquals(2, postfix.getBorderSize().getValue(), 0);
```
..and this one

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java:
```diff
@@ -342,49 +345,45 @@ public void test_diamond() throws IOException {
-        assertDMNStyle("Monospaced", 32, 3.5, 55, 66, 77, mydecision.getFontStyle());
+        assertColor(255, 255, 255, ((DMNStyle) mydecision.getStyle()).getFillColor());
+        assertColor(0, 0, 0, ((DMNStyle) mydecision.getStyle()).getStrokeColor());
+        //        assertEquals(0.5, mydecision.getBorderSize().getValue(), 0);
```
..and this one

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java:
```diff
@@ -342,49 +345,45 @@ public void test_diamond() throws IOException {
+        assertEquals(fontName, style.getFontFamily());
         assertEquals(fontSize, style.getFontSize(), 0);
-        assertEquals(fontBorderSize, style.getFontBorderSize(), 0);
+        //        assertEquals(fontBorderSize, style.getFontBorderSize(), 0);
```
..and, finally, this one :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/DMNMarshallerTest.java:
```diff
@@ -401,8 +400,12 @@ private static void assertColor(int r, int g, int b, Color color) {
+                   .filter(DMNShape.class::isInstance)
+                   .map(DMNShape.class::cast)
+                   .filter(shape -> shape.getDmnElementRef().getLocalPart().equals(id))
+                   .findFirst().get();
```
When we have a `.get` with no `isPresent` check, I think that's a good idea to explain why the value will always be present and do something like this:
```java
    private DMNShape findShapeByDMNI(org.kie.dmn.model.api.dmndi.DMNDiagram root, String id) {
        return root.getDMNDiagramElement()
                   .stream()
                   .filter(DMNShape.class::isInstance)
                   .map(DMNShape.class::cast)
                   .filter(shape -> shape.getDmnElementRef().getLocalPart().equals(id))
                   .findFirst()
                   .orElseThrow(() -> new UnsupportedOperationException("the explanation here"));
    }
```

tarilabs replied:
This is a test utility method, so I believe it would be straight-forward to understand any error, but anyway addressing as requested.

tarilabs commented:
For some reason this github page is not seeing https://github.com/tarilabs/kie-wb-common/commit/e839ba1ce7bfd5d04a78c60562993f813fbb0a11

tarilabs commented:
I'll try rebase in the hope this github page will notice https://github.com/tarilabs/kie-wb-common/commit/e839ba1ce7bfd5d04a78c60562993f813fbb0a11 has been added.

jomarko commented:
@tarilabs found one more issue during PR review [DROOLS-3185](https://issues.jboss.org/browse/DROOLS-3185)

tarilabs commented:
Okay I believe maybe I got where the problem is:
```
[INFO] 	at org.kie.workbench.common.dmn.backend.definition.v1_1.dd.PointUtils.point2dToDMNDIPoint(PointUtils.java:27)
[INFO] 	at org.kie.workbench.common.dmn.backend.DMNMarshaller.marshall(DMNMarshaller.java:472)
```

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -390,9 +392,25 @@ private void setConnectionMagnets(final Edge edge, String dmnEdgeElementRef, org
+                double ySource = yOfBound(upperLeftBound(sourceView));
+                connectionContent.setSourceConnection(MagnetConnection.Builder.at(source.getX() - xSource, source.getY() - ySource)); // Stunner connection x,y is relative to shape
+            } else { // fallback:
+                connectionContent.setSourceConnection(MagnetConnection.Builder.at(source.getX(), source.getY()));
```
I think this branch is not covered by tests.

tarilabs replied:
indeed because a fallback, as explained by @manstis this is a workaround to a Stunner issue

karreiro replied:
@tarilabs If this line does not deserve to be tested, could we remove it?

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -390,9 +392,25 @@ private void setConnectionMagnets(final Edge edge, String dmnEdgeElementRef, org
+                double yTarget = yOfBound(upperLeftBound(targetView));
+                connectionContent.setTargetConnection(MagnetConnection.Builder.at(target.getX() - xTarget, source.getY() - yTarget)); // Stunner connection x,y is relative to shape
+            } else { // fallback:
+                connectionContent.setTargetConnection(MagnetConnection.Builder.at(target.getX(), target.getY()));
```
..and this one.

tarilabs replied:
ref reply

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -464,16 +482,48 @@ public String marshall(final Diagram<Graph, Metadata> diagram) throws IOExceptio
+                            Point2D sourcePoint = ((Connection) connectionContent.getSourceConnection().get()).getLocation();
+                            Point2D targetPoint = ((Connection) connectionContent.getTargetConnection().get()).getLocation();
+                            if (sourcePoint == null) { // If the "connection source/target location is null" assume it's the centre of the shape.
+                                View<?> sourceView = (View<?>) e.getSourceNode().getContent();
```
This whole branch is not covered by tests.

tarilabs replied:
Disagree. This is the branch covered by the functional test. I would be happy to spare @jomarko manual functional test if there was a way.

jomarko replied:
Not sure what functional test do you mean @tarilabs?

karreiro replied:
I just ran all unit tests with a coverage tool and this branch continues uncovered.

tarilabs replied:
> I just ran all unit tests with a coverage tool and this branch continues uncovered.

@karreiro I believe I was not able to explain myself.

I have _no_ way to cover with a programmatic test this branch.

This branch is actually covered when @jomarko performs the tests manually, by hand.

This branch cover the requirements as explained by @manstis by the way Stunner framework works, rather its current limitation as explained in https://issues.jboss.org/browse/DROOLS-3184?focusedCommentId=13652929&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13652929, when creating a Diagram for DMN.

Insisting of running a code-coverage tool won't sort any effect, until there is a way to make a DMN Diagram in a programmatic way based on the Stunner canvas.
I was told when originally providing the marshaller that there is no feasible way to automate the test which today @jomarko is doing manually, by hand.

That is the reason why test in the codebase starts with a DMNDiagram in xml form, and de-serialize this into a Stunner Diagram, to cycle in the Marshaller, and finally perform assertion; please notice all test case start from a DMN xml for this reason.

So unless there is a way now to programmatically instrument Stunner to script it like "Create a Decision shape, click on constellation button other-decision, drag shape to this location x,y, etc." there will be no way to cover this **specific** branch from a purely code perspective. As this branch is meant to cover the Stunner limitation from link above, so it would occur only and only when the Diagram is created directly from Stunner.

Was I able to fully explain this time why code-coverage metric is not relevant for this case?

Naturally when there are no Stunner limitations, I was able to cover anyway with cycling the model via the marshaller.

manstis replied:
We can create Stunners graph programmatically to represent what would have been created manually. No problem. I tend to agree with Guilherme that we _could_ write unit tests to cover the code branches.

tarilabs replied:
Cool ! 
So this is new since the original test case I built upon the current test base.
So I am missing the script/list of API calls which performs the click on the constellation button, along with the basic one to instrument in order to "click on Decision", "Drag here" etc.
Is there a test case I can copy, please?

tarilabs replied:
(by "copy" I don't mean I expect a test case to be there already providing all I need, but something to take at least "inspiration" from ;) )

manstis replied:
Heh, I didn't say there was a builder API or scripting facility... but what we can definitely build rather manually is the Stunner 'Graph' object with Nodes and Edges representing what can be constructed using the UI. Definitely a bunch of boiler plate but definitely doable.

manstis replied:
I can help (on Monday) for sure.. I think @jomarko may have some tests already that go from 'Graph' to XML and back again.. I'll need to logon to my laptop.. on Monday!

tarilabs replied:
:(
It was sounding too good to be true :(

So this confirms there is no realistic way to automate the manual test.

So this means to cover this branch I will have to re-create manually hand-crafted/hard-coded the same behavior of the null limitation as explained in the linked JIRA.

If it's really a necessity I can do, but I would suggest the title should have a very well written documentation (in-code) that once Stunner limitation is overcome, the test ought to be removed or might need again manual adjustment.

Please let me know if this code-coverage metric is a strict requirement, given is to cover an edge-case (of the Stunner limitation of the null Point2D)

manstis replied:
... I have 2 PRs waiting for this to be merged.. so it's in my interest to get this one in

tarilabs replied:
ok

manstis replied:
Is it an edge case?!?! It's easily produced from the UI so we have to expect users to create the scenario... and so should probably be tested.

tarilabs replied:
> Is it an edge case?
On linked JIRA:
> Undefined i.e. null (IDK why, need to raise an issue with the Stunner people)
My understanding this is an issue with Stunner currently using the constellation button rather than using the palette. I have assumed then this will be solved from Stunner framework eventually to align with the palette behavior.
But I may have simply misunderstood the explaination then?

tarilabs replied:
covered.

manstis replied:
@tarilabs Thank-you for taking the time for the new test.

tarilabs replied:
> for the new test.

No problem, thank you to you @manstis for explaining which strategy to pursue in order to complete code-coverage.

There is still some things which I don't understand about the framework, but I would suggest best to also discuss them at the first chance we have a call.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java:
```diff
@@ -464,16 +482,48 @@ public String marshall(final Diagram<Graph, Metadata> diagram) throws IOExceptio
+                                sourcePoint = Point2D.create(xSource + sourcePoint.getX(), ySource + sourcePoint.getY());
+                            }
+                            if (targetPoint == null) { // If the "connection source/target location is null" assume it's the centre of the shape.
+                                double xTarget = xOfBound(upperLeftBound(view));
```
This branch is not covered too.

tarilabs replied:
ref reply.

tarilabs replied:
covered.

tarilabs commented:
I have added an additional commit which resolve an issue as the unmarshalling code needed to be transformed into a 2-pass, as Stunner rely on relative positioning for the connection of the Edges.
Then I've done the following.
I included 1 InputData and 1Decision and an InformationRequirement edge.
I have saved and reopened multiple time.
Between re-open, before saving again, I changed the location of the "magnet" to be either of the center, left, right, etc. and mixing the same on the 2 Shapes/Nodes.
I done at least 1 combination of each different, and once Saved and re-opened, it was preserved where I did expect it.
Finally, I did the magnet like this:
![image](https://user-images.githubusercontent.com/1699252/47574740-8f4b2100-d940-11e8-91d3-c11867520d32.png)
and I checked the XML was consistent.
I also checked on the Partner editor:
![image](https://user-images.githubusercontent.com/1699252/47574764-a12cc400-d940-11e8-8048-49a8781932be.png)

@jomarko please notice that the Partner editor make a change if the edge crosses the shape, to avoid the crossing --or at least this is what I noticed.
For example, if you do on our editor something like:
![image](https://user-images.githubusercontent.com/1699252/47574854-e3ee9c00-d940-11e8-876b-109c060037fc.png)
the XML will be correct:
![image](https://user-images.githubusercontent.com/1699252/47574910-07b1e200-d941-11e8-97a7-1a7dca6aa450.png)
but on the Partner editor it will render as:
![image](https://user-images.githubusercontent.com/1699252/47574971-30d27280-d941-11e8-98d8-a3f08817d0f8.png)

tarilabs commented:
Functional test as a synonym of manual, non-programmable test, executed by human not the machine

jomarko commented:
@tarilabs would you mind to use Logger instead of `System.out.println`? I am not going to do detail re-review. Check just few cases of diagrams where magnet points were used.
